consensus/propagation/reactor.go

```diff
@@ -135,3 +158,255 @@ func (blockProp *Reactor) Receive(chID byte, peer p2p.Peer, msgBytes []byte) {
+}
+
+// broadcastSelfProposalHaves broadcasts the haves to all the connected peers when we're the proposers.
+// Note: the haves are chunked so that every peer only receives a portion of the haves.
+//
+//nolint:unused
+func (blockProp *Reactor) broadcastSelfProposalHaves(proposal *types.Proposal, from p2p.ID, haves *bits.BitArray) {
+}
+
+// chunkParts takes a bit array then returns an array of chunked bit arrays.
+// TODO document how the redundancy and the peer count are used here.
+func chunkParts(p *bits.BitArray, peerCount, redundancy int) []*bits.BitArray {
+	size := p.Size()
+	if peerCount == 0 {
+		peerCount = 1
+	}
+	chunkSize := size / peerCount
+	// round up to use the ceil
+	if size%peerCount != 0 || chunkSize == 0 {
+		chunkSize++
+	}
+
+	// Create empty bit arrays for each peer
+	parts := make([]*bits.BitArray, peerCount)
+	for i := 0; i < peerCount; i++ {
+		parts[i] = bits.NewBitArray(size)
+	}
+
+	chunks := chunkIndexes(size, chunkSize)
+	cursor := 0
+	for p := 0; p < peerCount; p++ {
+		for r := 0; r < redundancy; r++ {
+			start, end := chunks[cursor][0], chunks[cursor][1]
+			for i := start; i < end; i++ {
+				parts[p].SetIndex(i, true)
+			}
+			cursor++
+			if cursor >= len(chunks) {
+				cursor = 0
+			}
+		}
+	}
+
+	return parts
+}
+
+// chunkIndexes
+// TODO document and explain the parameters
+func chunkIndexes(totalSize, chunkSize int) [][2]int {
+	if totalSize <= 0 || chunkSize <= 0 {
+		panic(fmt.Sprintf("invalid input: totalSize=%d, chunkSize=%d \n", totalSize, chunkSize))
+		// return nil // Handle invalid input gracefully
+	}
+
+	var chunks [][2]int
+	for start := 0; start < totalSize; start += chunkSize {
+		end := start + chunkSize
+		if end > totalSize {
+			end = totalSize // Ensure the last chunk doesn't exceed the total size
+		}
+		chunks = append(chunks, [2]int{start, end})
+	}
+
+	return chunks
+}
+
+// handleHaves is called when a peer sends a have message. This is used to
+// determine if the sender has or is getting portions of the proposal that this
+// node doesn't have. If the sender has parts that this node doesn't have, this
+// node will request those parts.
+// The peer must always send the proposal before sending parts. If they did
+// not, this node must disconnect from them.
+// This method will:
+// - get the provided peer from the peer state
+// - get the proposal referenced in the haves message
+// - set the provided haves as the peer's haves
+// - if the returned parts from the proposal are complete, we return
+// - otherwise, we check if the sender has parts that we don't have.
+// - if they do, we check if we already requested those parts enough times (a limit will be defined)
+// of we already requested the parts from them.
+// - if so, we just gossip the haves to our connected peers.
+// - otherwise, we send the wants for the missing parts to that peer before broadcasting the haves.
+// - finally, we keep track of the want requests in the proposal state.
```

Comment:
this is too low level therefore difficult to parse and determine if its sufficient without seeing the actual code

upon seeing a have, the node must  
1) determine if this have has already been seen. If so, ignore.
2) check the validity of the message and the proposal for that height and round has been received already. If invalid or the proposal is unseen, disconnect from the peer and return
3) if we don't have any of the parts that are in this have, request them.
4) if its the first time that we have seen haves for these parts, broadcast the message to all peers (not the sender ofc)